### ISSUE TITLE
Lint for `use ident;`

### ISSUE THREAD

**Issue by jhpratt** [rust-lang/rust-clippy#4903]
Any statement of the form `use ident;` is inherently useless, as it brings nothing new into scope. It can safely be removed in all situations (unless there's an edge case I'm not aware of?).

**Comment by flip1995** [comment]
This is useful to only import crate scopes:

```rust
//! lib.rs of crate1
fn foo() {}
```

```rust
//! lib.rs of crate2 
fn foo() {}
```

So crate1 and crate2 have 2 functions with the same name. Now it is impossible to import both `foo` functions at the same time directly:

```rust
use crate1::foo;
use crate2::foo;

fn main() {
    foo(); // which `foo` is used here?
}
```

So what you want to do in this situation is

```rust
use crate1;
use crate2;

fn main() {
    crate1::foo();
    crate2::foo();
}
```

---
Something similar is done in clippy, when both [`hir::Expr`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc/hir/struct.Expr.html) and [`ast::Expr`](https://doc.rust-lang.org/nightly/nightly-rustc/syntax/ast/struct.Expr.html) is required in a module.

**Comment by jhpratt** [comment]
@flip1995 That would only be the case in Rust 2015, correct?

**Comment by flip1995** [comment]
No this is also the case in rust 2018, like in my example above, which is only valid in the 2018 edition.

**Comment by jhpratt** [comment]
I think you might not understand exactly what I'm saying. Importing paths is, of course, extremely useful. I agree it's not always feasible to do so. However, in Rust 2018 at least, you could eliminate the `use` statements from the latter snippet with zero offect whatsoever. This has nothing at all to do with naming conflicts.

**Comment by flip1995** [comment]
> you could eliminate the use statements from the latter snippet with zero offect whatsoever.

Ahhh, now it makes sense. Thanks for clarifying!
